blog/2024-01-12-implement-async-storage-data-streaming-pytorch/index.md

```diff
@@ -0,0 +1,256 @@
+description: Discover how to implement async storage data streaming in PyTorch projects. This blog provides a practical approach to setting up a custom PyTorch's IterableDataset that reads data from a time series database.
+authors: anthony
+tags: [AI, DataStreaming, PyTorch]
+slug: ai/datastreaming/pytorch/implement-async-storage-data-streaming-pytorch
+date: 2024-01-12
+image: ./img/pytorch-iterabledataset.webp
+---
+
+![PyTorch Training Diagram](./img/pytorch-iterabledataset.webp)
+<small>PyTorch training loop with data streaming from remote device</small>
+
+When training a model, we aim to process data in batches, shuffle data at each epoch to avoid over fitting, and leverage Python's `multiprocessing` for data fetching through multiple workers.
+
+The reason that we want to use multiple workers is that GPUs are capable of handling large amounts of data concurrently; however, the bottleneck often lies in the time-consuming task of loading this data into the system.
+
+Moreover, the challenge is even trickier when there is simply too much data to store the whole dataset on disk and we need to stream data from an async storage database such as [**ReductStore**](<https://www.reduct.store/>).
+
+Addressing this challenge requires a custom Python iterator to bridge asynchronous data fetching with a synchronous [**PyTorch's IterableDataset**](<https://pytorch.org/docs/stable/data.html#torch.utils.data.IterableDataset>).
+
+In this blog post, we will go through a full example and setup a data stream to [**PyTorch**](<https://pytorch.org/>) from a playground dataset on a remote database.
+
+Let's dig in!
+
+<!--truncate-->
+
+## Setting Up the Python Client for ReductStore Database
+To use ReductStore's playground dataset, follow these steps to get started with [**Python's client**](<https://py.reduct.store/en/latest/>):
+
+- Install the `reduct` Python package using pip: `pip install reduct-py`.
+
+- Import the Client class from the `reduct` package at the beginning of your script: `from reduct import Client`.
+
+- Define your connection parameters:
+
+    - **HOST**: The URL of your ReductStore instance (e.g., `"https://play.reduct.store"`).
+
+    - **API\_TOKEN**: Your API token for authentication (e.g., `"dataset-read-eab13e4f5f2df1e64363806443eea7ba83406ce701d49378d2f54cfbf02850f5"`).
+
+    - **BUCKET**: Name of the bucket containing your data (e.g., `"datasets"`).
+
+    - **DATASET**: Specific dataset you want to access (e.g., `"cats"`).
+
+    <!-- -->
+
+- Initialize the client:
+
+
+<!-- -->
+
+```python
+client = Client(HOST, api_token=API_TOKEN)
+```
+
+### Getting data from playground server
+To use the 'Cats' dataset with Python, you'll need to follow a few other straightforward steps.
+
+First, ensure you have installed necessary libraries like NumPy and OpenCV.
+
+Then access your dataset by asynchronously querying records within a specified range using, e.g. `bucket.query(DATASET, start=0, stop=5)`.
```

Comment:
here I would explain that ReductStore is a time-series database and we usually use timestamps to specify the range. But the timestamps are a number of microseconds from 1970-01-01, so we can use them as IDs of records.